### ISSUE TITLE
cancelled events are not evicted from batch

### ISSUE THREAD

**Issue by talevy** [elastic/logstash#6055]
Dropping events using `drop {}` does not actually drop events

recreate using:

```
input {
    generator {
        count => 1
        message => "hello"
    }
}

filter {
    drop {}
    mutate { add_tag => ["hello"] }
}

output {
    stdout { codec => rubydebug }
}

```

following change fixes this, but makes things a little less efficient

```ruby
diff --git a/logstash-core/lib/logstash/filters/base.rb b/logstash-core/lib/logstash/filters/base.rb
index 35bf49e..310d87b 100644
--- a/logstash-core/lib/logstash/filters/base.rb
+++ b/logstash-core/lib/logstash/filters/base.rb
@@ -151,10 +151,10 @@ class LogStash::Filters::Base < LogStash::Plugin
     LogStash::Util.set_thread_plugin(self)
     result = []
     events.each do |event|
-      unless event.cancelled?
         result << event
-        filter(event){|new_event| result << new_event}
-      end
+        unless event.cancelled?
+          filter(event){|new_event| result << new_event}
+        end
     end
     result
   end
```

**Comment by colinsurprenant** [comment]
This is not the proper fix. It solves as a side effect. Working on a PR.

**Comment by colinsurprenant** [comment]
proposed fix in #6059

**Comment by colinsurprenant** [comment]
fixed by #6059
